pyogrio/_io.pyx

```diff
@@ -1142,7 +1143,7 @@ def ogr_read_arrow(
 
         # stream has to be consumed before the Dataset is closed
         import pyarrow as pa
-        table = pa.RecordBatchStreamReader._import_from_c(stream_ptr).read_all()
+        reader = pa.RecordBatchStreamReader._import_from_c(stream_ptr)
```

Comment:
I think you need to define `reader = None` before the `try` block, and then check against that before you close it:

```
if reader is not None:
    reader.close()
```


Otherwise, we get nested errors like this:

```
>>> t = read_arrow('foo')
Traceback (most recent call last):
  File "pyogrio/_io.pyx", line 150, in pyogrio._io.ogr_open
    ogr_dataset = exc_wrap_pointer(
  File "pyogrio/_err.pyx", line 178, in pyogrio._err.exc_wrap_pointer
    raise exc
pyogrio._err.CPLE_OpenFailedError: foo: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pyogrio/_io.pyx", line 160, in pyogrio._io.ogr_open
    raise DataSourceError(str(exc))
pyogrio.errors.DataSourceError: foo: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bcward/projects/pyogrio/pyogrio/raw.py", line 199, in read_arrow
    with open_arrow(
  File "/Users/bcward/.pyenv/versions/3.10.5/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "pyogrio/_io.pyx", line 1162, in ogr_open_arrow
    reader.close()
UnboundLocalError: local variable 'reader' referenced before assignment
```